flight/Modules/Logging/logging.c

```diff
@@ -381,25 +353,135 @@ static void loggingTask(void *parameters)
 				read_sector = loggingData.FileSectorNum;
 			}
 			LoggingStatsSet(&loggingData);
-
+			// fall-through to default case
+		default:
+			//  Makes sure that we are not hogging the processor
+			PIOS_Thread_Sleep(10);
+			// Close the file if necessary
+			if (write_open) {
+				PIOS_STREAMFS_Close(streamfs_id);
+				loggingData.MinFileId = PIOS_STREAMFS_MinFileId(streamfs_id);
+				loggingData.MaxFileId = PIOS_STREAMFS_MaxFileId(streamfs_id);
+				LoggingStatsSet(&loggingData);
+				write_open = false;
+			}
 		}
+	}
+}
 
-		i++;
+/**
+ * Forward data from UAVTalk out the serial port
+ * \param[in] data Data buffer to send
+ * \param[in] length Length of buffer
+ * \return -1 on failure
+ * \return number of bytes transmitted on success
+ */
+static int32_t send_data(uint8_t *data, int32_t length)
+{
+	if( PIOS_COM_SendBuffer(logging_com_id, data, length) < 0)
+		return -1;
+
+	written_bytes += length;
+
+	return length;
+}
+
+
+/**
+ * Function for sorting linked list based on logging period and makes
+ * sure that objects with private queues come first.
+ */
+static int info_cmp(struct UAVOLogInfo *info_1, struct UAVOLogInfo *info_2)
+{
+	if ((info_1->queue == shared_queue) && (info_2->queue != shared_queue)){
+		return true;
+	}
+	if ((info_1->queue != shared_queue) && (info_2->queue == shared_queue)){
+		return false;
 	}
+
+	return info_1->logging_period > info_2->logging_period;
 }
 
+/**
+ * @brief Callback for adding an object to the logging queue
+ * @param ev the event
+ */
+static void obj_updated_callback(UAVObjEvent * ev)
+{
+	struct UAVOLogInfo *info;
+	uint32_t now;
+
+	if (loggingData.Operation != LOGGINGSTATS_OPERATION_LOGGING){
+		// We are not logging, so all events are discarded
+		return;
+	}
+	LL_FOREACH(log_info, info) {
```

Comment:
Hmm.. you are right, the scenario where a high frequency object has a low logging rate will indeed increase the CPU utilization. Which isn't "ideal".

Your suggestion to implement this in the uavobjectmanager callback layer makes a lot of sense. I have actually been thinking about this as well. We could have a `UAVObjConnectThrottledCallback` function that allows passing a minimum interval at which the callback is called. This could be useful for things like telemetry as well. The downside would be that we would need to extend `struct ObjectEventEntry` to include the interval which would increase the memory requirements by 4 bytes (assuming the interval is `uint32`), even for targets that don't use logging.